src/Moq/ExpressionExtensions.cs

```diff
@@ -324,8 +324,10 @@ internal static PropertyInfo GetReboundProperty(this MemberExpression expression
 					.GetMember(property.Name, MemberTypes.Property, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)
 					.Cast<PropertyInfo>()
 					.SingleOrDefault(p => p.PropertyType == property.PropertyType);
-				if (derivedProperty != null && derivedProperty.GetMethod.GetBaseDefinition() == property.GetMethod)
+				if (derivedProperty != null)
 				{
+					if ((derivedProperty.CanRead && derivedProperty.GetGetMethod(true).GetBaseDefinition() == property.GetGetMethod(true)) ||
+						(derivedProperty.CanWrite && derivedProperty.GetSetMethod(true).GetBaseDefinition() == property.GetSetMethod(true)))
```

Comment:
This is what causes the `NullReferenceException`. When upgrading from a base property to a more derived one, the more derived one may not have the same accessors as the base one did.